dmipy/optimizers/amico_cvxpy.py

```diff
@@ -0,0 +1,143 @@
+
+        # Compute length of the vector x0
+        x0_len = 0
+        for m_idx in range(self.model.N_models):
+            m_atoms = 1
+            for p in self.model.models[m_idx].parameter_names:
+                if self.model.model_names[m_idx] + p in grid_params:
+                    m_atoms *= Nt
+            x0_len += m_atoms
+
+        # Creating parameter tessellation grids and corresponding indices
+        self.grids, self.idx = {}, {}
+        for m_idx in range(self.model.N_models):
+            model = self.model.models[m_idx]
+            model_name = self.model.model_names[m_idx]
+
+            param_sampling, grid_params_names = [], []
+            m_atoms = 1
+            for p in model.parameter_names:
+                if model_name + p not in grid_params:
+                    continue
+                grid_params_names.append(model_name + p)
+                p_range = self.model.parameter_ranges[model_name + p]
+                self.grids[model_name + p] = np.mean(p_range)
+                param_sampling.append(np.linspace(p_range[0], p_range[1],
+                                                  self.Nt, endpoint=True))
+                m_atoms *= self.Nt
+
+            self.idx[model_name] =\
+                sum([len(self.idx[k]) for k in self.idx]) + np.arange(0, m_atoms)
+            params_mesh = np.meshgrid(*param_sampling)
+            for p_idx, p in enumerate(grid_params_names):
+                self.grids[p] += np.zeros(x0_len, dtype=np.float32)
+                self.grids[p][self.idx[model_name]] = np.ravel(params_mesh[p_idx])
+
+            self.grids['partial_volume_' + str(m_idx)] = np.zeros(x0_len)
+            self.grids['partial_volume_' + str(m_idx)][self.idx[model_name]] = 1.
+
+        arguments = self.grids.copy()
+        arguments[dir_params[0]] = [0, 0]
+        self.M = self.model.simulate_signal(acquisition_scheme, arguments)
+        self.M = self.M[:, ~acquisition_scheme.b0_mask]
+
+    def __call__(self, data):
+        """
+        The fitting function of AMICO optimizer.
+        Parameters
+        ----------
+        data : Array of size (Ndata)
+            The normalized dMRI signal attenuation to be fitted.
+
+        Returns
+        -------
+        fitted_parameter_vector : Array of size (Nx),
+            Vector containing probability distributions of the parameters that are
+            being estimated
+        """
+
+        x0 = cvxpy.Variable(len(self.x0_vector))
+
+        cost = 0.5 * cvxpy.sum_squares(self.M * x0 - data)
+        for m_idx, model_name in enumerate(self.model.model_names):
+            cost += self.lambda_1[m_idx] *\
+                    cvxpy.norm(x0[self.idx[model_name]], 1)
+            cost += 0.5 * self.lambda_2[m_idx] *\
+                    cvxpy.norm(x0[self.idx[model_name]], 2) ** 2
+
+        problem = cvxpy.Problem(cvxpy.Minimize(cost), [x0 >= 0])
+        problem.solve()
+
```

Comment:
I suggest adding a TODO here for that we may want to do the extra M-pruning and l1-free optimization if an l1-norm was added (what you correctly suggested in our last dmipy meeting).

Reply:
Ok, I will add it.